### PR TITLE
feat: Performance API: Transaction

### PR DESCRIPTION
**Pull request by brustolin**
## :scroll: Description

Created a folder with performance API.
Hacked the hell out of the SDK to be able to send a transaction to Sentry.

image

image

image

## :bulb: Motivation and Context

Add performance API to sentry-cocoa SDK.

## :green_heart: How did you test it?

This is a draft PR, trying it out from the Sentry iOS app, checking result in Sentry.
Test will came later.

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the CHANGELOG
- [ ] I updated the docs if needed
- [ ] Review from the native team if needed
- [x] No breaking changes

## :crystal_ball: Next steps

TBD

### CODE REVIEW

File: Samples/iOS-ObjectiveC/iOS-ObjectiveC/ViewController.m
Comment (bruno-garcia) [inline Samples/iOS-ObjectiveC/iOS-ObjectiveC/ViewController.m:113]: once we get `captureTransaction` to `SentryHub`, we can get rid of this, and on `[fakeTransaction finish]` we'll capture it to Sentry

File: Sources/Sentry/Transactions/ISentrySpan.h
Comment (bruno-garcia) [inline Sources/Sentry/Transactions/ISentrySpan.h:8]:
```suggestion
```

File: Sources/Sentry/Transactions/SentrySpanContext.m
Comment (bruno-garcia) [inline Sources/Sentry/Transactions/SentrySpanContext.m:8]:
```suggestion
```

File: Sources/Sentry/Transactions/SentryTransaction.h
Comment (bruno-garcia) [inline Sources/Sentry/Transactions/SentryTransaction.h:8]:
```suggestion
```

**Comment by bruno-garcia** [comment]
We should agree on the next steps. I guess if we can `startTransaction` on the `Hub` and then have `finish` on the transaction itself capture it, that could be a good milestone. So build from here to do that, plus get tests in so we can merge.

Probably best to keep a branch to merge into, and we pull `master` into that branch from time to time. So that we can continue to release `master` without having to add the incomplete performance feature.

@philipphofmann @marandaneto wdyt?

**Comment by bruno-garcia** [comment]
I changed the `base` branch to `feat/performance-monitoring`, which can be used for all PRs related to performance. Draft PR for that branch for visibility: https://github.com/getsentry/sentry-cocoa/pull/909

File: Sources/Sentry/Transactions/ISentrySpan.h
Comment (marandaneto) [reply Sources/Sentry/Transactions/ISentrySpan.h:8]: this could be configured on Xcode, so it does not add any more.

**Comment by marandaneto** [comment]
> We should agree on the next steps. I guess if we can `startTransaction` on the `Hub` and then have `finish` on the transaction itself capture it, that could be a good milestone. So build from here to do that, plus get tests in so we can merge.
> 
> Probably best to keep a branch to merge into, and we pull `master` into that branch from time to time. So that we can continue to release `master` without having to add the incomplete performance feature.
> 
> @philipphofmann @marandaneto wdyt?

that would be ideal

File: Sources/Sentry/Transactions/SentrySpanContext.h
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentrySpanContext.h:31]:
```suggestion
- (instancetype)initWithTraceId:(SentryId *)traceId
                         spanId:(SentrySpanId *)spanId
```

File: Sources/Sentry/Transactions/SentrySpanContext.m
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentrySpanContext.m:27]: `l`: We can simplify this a bit
```suggestion
    return [self initWithSampled:false];
```

File: Sources/Sentry/Transactions/SentrySpanContext.m
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentrySpanContext.m:55]: `l`: Why not initialize the dict in the constructor so we can remove the check for `nil` in here.

File: Sources/Sentry/Transactions/SentryTransaction.h
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentryTransaction.h:61]:
```suggestion
```

File: Sources/Sentry/Transactions/SentrySpanContext.m
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentrySpanContext.m:16]: `l`: I think we could remove this and also `- (NSMutableDictionary *)tags`. We could just allocate the dictionary on line 47, or am I missing something?

File: Sources/Sentry/Transactions/SentryTransaction.h
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentryTransaction.h:65]: `l`:
```suggestion
                     spanContext:(SentrySpanContext *)spanContext
```

File: Sources/Sentry/Transactions/SentryTransaction.h
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentryTransaction.h:62]: `l`:
```suggestion
- (instancetype)initWithTransactionContext:(SentryTransactionContext *)transactionContext
```

File: Sources/Sentry/Transactions/SentryTransaction.m
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentryTransaction.m:34]: `l`:
```suggestion
    NSMutableDictionary<NSString *, id> *serializedData = @{
```

File: Sources/Sentry/Transactions/SentryTransaction.m
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentryTransaction.m:37]: Why is this empty?

File: Sources/Sentry/Transactions/SentryTransaction.m
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentryTransaction.m:91]:
```suggestion
```

File: Sources/Sentry/Transactions/SentryTransaction.h
Comment (philipphofmann) [inline Sources/Sentry/Transactions/SentryTransaction.h:15]: `h`: Please add `NS_SWIFT_NAME`

**Review by philipphofmann — COMMENTED** [review]
Very nice hacking. I'm excited to see some transactions in our SDK. 🥳 

I know this is a draft PR, but still, I made lots of comments, so we can get this PR ready for review.

Please move all public headers into the `Sources/Public` folder see https://github.com/getsentry/sentry-cocoa/blob/master/CONTRIBUTING.md#public-headers. Please also move all `.m` files to `Sources/Sentry`.

I'm looking forward to seeing some tests for all of this.

File: Sources/Sentry/Transactions/SentrySpanId.h
Comment (marandaneto) [reply Sources/Sentry/Transactions/SentrySpanId.h:12]: they are different, See java https://github.com/getsentry/sentry-java/blob/3c87a48442e536a41ed905d9e43bdc2f7b800176/sentry/src/main/java/io/sentry/SpanId.java#L20-L22
only 16 chars

File: Sources/Sentry/Transactions/SentrySpanContext.m
Comment (brustolin) [reply Sources/Sentry/Transactions/SentrySpanContext.m:55]: This part was copied from SentryEvent. I guess we could.

File: Sources/Sentry/Transactions/SentryTransaction.m
Comment (brustolin) [reply Sources/Sentry/Transactions/SentryTransaction.m:37]: Bruno wanted to see some progress so I created this draft PR. Still work to do before consider it done.

**Comment by marandaneto** [comment]
@brustolin just to explain the comments prefixes:

`l`: low, that means, its not important for the reviewer, you could ignore but it'd be nice to be fixed.
`m`: mid, it's sort of nice that you both agree either on a solution or to ignore.
`h`; high, its really important for the reviewer, you definitely should look into it.

File: Sources/Sentry/Transactions/ISentrySpan.h
Comment (bruno-garcia) [reply Sources/Sentry/Transactions/ISentrySpan.h:13]: @philipphofmann what will we name the actual class? On Android we chose to use `I` as a convention for such cases where the interface matches the implementation name .. Actually I think we just endded up prefixing all with `I` at the end, similar to C#

File: Sources/Sentry/Transactions/SentrySpanContext.m
Comment (bruno-garcia) [reply Sources/Sentry/Transactions/SentrySpanContext.m:55]: Initializing in the constructor only we're certain this will always be used otherwise we're just allocating for no use.

File: Sources/Sentry/Transactions/ISentrySpan.h
Comment (philipphofmann) [reply Sources/Sentry/Transactions/ISentrySpan.h:13]: I know the `I` pattern, and personally, I'm not a huge fan of it. I think you should avoid it whenever possible. Name the interface after the concept and the actual implementation should tell more about the implementation itself. For example, in Java, you have `List` and `ArrayList`, `CopyOnWriteArrayList`, ...
The `I` pattern is more common in Java and C#, but I don't think it is common in Objective-C or Swift, or am I wrong @brustolin?

File: Sources/Sentry/Transactions/SentrySpanId.h
Comment (philipphofmann) [reply Sources/Sentry/Transactions/SentrySpanId.h:12]: Ah, thanks for pointing that out. The comment says 36 chars. It would be nice to update the comments then @brustolin 😀.

File: Sources/Sentry/Transactions/SentrySpanContext.m
Comment (philipphofmann) [reply Sources/Sentry/Transactions/SentrySpanContext.m:55]: I didn't think about that @bruno-garcia, so I'm fine with keeping it then.

File: Sources/Sentry/Transactions/SentryTransaction.h
Comment (bruno-garcia) [reply Sources/Sentry/Transactions/SentryTransaction.h:15]: What would the `NS_SWIFT_NAME` be? Just `Transaction`?
That's a very broad name, can we keep it SentryTransaction though?

File: Sources/Sentry/SentryEnvelope.m
Comment (brustolin) [inline Sources/Sentry/SentryEnvelope.m]: Wondering whether this change can cause some errors.

**Comment by brustolin** [comment]
I think I fixed everything you guys commented.

By Bruno request I changed SentryTransaction inheritance to SentryEvent and remove some of the previous SentryTransaction handlers like "sendTransaction" from SentryClient to use "sendEvent", this way the client can prepare the envelope with all required information.

SentryTransaction tests included

File: Sources/Sentry/Public/SentryClient.h
Comment (marandaneto) [inline Sources/Sentry/Public/SentryClient.h:44]: missing the `withScope`

File: Sources/Sentry/Public/SentryHub.h
Comment (marandaneto) [inline Sources/Sentry/Public/SentryHub.h:5]: on Java/.NET, we call it only `TransactionContext`

File: Sources/Sentry/SentryClient.m
Comment (marandaneto) [inline Sources/Sentry/SentryClient.m:237]: this could call either `captureEnvelope` or the transport directly, remember to append those values to the transaction like `release`, `environment`, `tags`, probably from SentryOptions

File: Sources/Sentry/Transactions/ISentrySpan.h
Comment (marandaneto) [reply Sources/Sentry/Transactions/ISentrySpan.h:5]: I guess they can have the same name, but the file has a different extension, using `@interface` and `@ implementation `, if this holds true, I'd avoid the `I` as well

File: Sources/Sentry/Transactions/SentrySpanContext.h
Comment (marandaneto) [inline Sources/Sentry/Transactions/SentrySpanContext.h:16]: we could call them `SpanId` only instead of `SentrySpanId`, Java/.NET is like this

File: Sources/Sentry/Transactions/SentrySpanContext.h
Comment (marandaneto) [inline Sources/Sentry/Transactions/SentrySpanContext.h:22]: why not only `description`

**Comment by marandaneto** [comment]
> I think I fixed everything you guys commented.
> 
> By Bruno request I changed SentryTransaction inheritance to SentryEvent and remove some of the previous SentryTransaction handlers like "sendTransaction" from SentryClient to use "sendEvent", this way the client can prepare the envelope with all required information.
> 
> SentryTransaction tests included

nice, thanks for that.

I've written a few more comments.
btw please resolve the comments that are already addressed, it's easier to review :)

File: Sources/Sentry/SentryClient.m
Comment (brustolin) [reply Sources/Sentry/SentryClient.m:237]: Thats the point. Capture event append all these values already, lets say the SDK grows and more properties are necessary, we would have to change in one place. Do you think the best option it is to make them part ways?

File: Sources/Sentry/Transactions/SentrySpanContext.h
Comment (brustolin) [reply Sources/Sentry/Transactions/SentrySpanContext.h:22]: description is a Objc NSObject method already. Like toString from java/.NET. I guess we could override it and create the setDescription, I dont know if  I like this option.

File: Sources/Sentry/SentryClient.m
Comment (marandaneto) [reply Sources/Sentry/SentryClient.m:237]: at least on Java/Android we've decided to not do it because transactions are supposed to be cheap, if we do captureEvent on Android it runs event processors, beforeSend etc, reads available memory, free storage, etc, it'd be very expensive and also having a bunch of data in the JSON that are not that useful, so we decided going forward adding the missing attributes later on (based on needs), I believe on iOS would be the same.

Comment (marandaneto) [reply Sources/Sentry/SentryClient.m:237]: you could still extract a method with the common attributes from captureEvent and captureTransaction and call that method, right, from both places.

File: Sources/Sentry/SentryEnvelope.m
Comment (brustolin) [reply Sources/Sentry/SentryEnvelope.m]: Code removed to use send event instead.

File: Sources/Sentry/Transactions/SentrySpanContext.h
Comment (marandaneto) [reply Sources/Sentry/Transactions/SentrySpanContext.h:22]: oh I see its a reserved word, well that's ok then I'd say, to keep as it is

File: Sources/Sentry/SentryEnvelope.m
Comment (brustolin) [reply Sources/Sentry/SentryEnvelope.m]: This code was removed. Using sendEvent instead

File: Sources/Sentry/SentryEnvelope.m
Comment (marandaneto) [reply Sources/Sentry/SentryEnvelope.m]: @philipphofmann

File: Sources/Sentry/Transactions/ISentrySpan.h
Comment (brustolin) [reply Sources/Sentry/Transactions/ISentrySpan.h:5]: Thats correct, the I pattern is not common in Objc. Since I was testing I just copied from java implementation until I realized I didn't need it to capture a transaction, so, some classes are not being used yet.

Open to new name options.

Apple naming suggestion: 
"Most protocols group related methods that aren’t associated with any class in particular. This type of protocol should be named so that the protocol won’t be confused with a class. A common convention is to use a gerund (“...ing”) form"

"Some protocols group a number of unrelated methods (rather than create several separate small protocols). These protocols tend to be associated with a class that is the principal expression of the protocol. In these cases, the convention is to give the protocol the same name as the class."

I guess we could go for the same name?

File: Sources/Sentry/Transactions/SentryTransaction.h
Comment (brustolin) [reply Sources/Sentry/Transactions/SentryTransaction.h]: I will use SentryTransaction as Java does.

File: Sources/Sentry/Public/SentryHub.h
Comment (brustolin) [reply Sources/Sentry/Public/SentryHub.h:5]: This was one of the first things I added, was trying to figure Objc SDK naming convention. I guess I can fallow Java naming correct?

Comment (marandaneto) [reply Sources/Sentry/Public/SentryHub.h:5]: mmm usually we just call as it is, we only add `Sentry` in front of what could bring import conflicts, so we run into issues like https://github.com/getsentry/sentry-dart/issues/250

yeah I'd prefer to keep rather the same whenever possible.

File: Sources/Sentry/Transactions/ISentrySpan.h
Comment (marandaneto) [reply Sources/Sentry/Transactions/ISentrySpan.h:5]: sounds good to me

File: Sources/Sentry/Public/SentryHub.h
Comment (bruno-garcia) [reply Sources/Sentry/Public/SentryHub.h:5]: In Objective-C it's even more important to avoid conflicts since it'll happen even outside of an `import` like in Dart. So I'd say prefixing with Sentry here is probably the best option

File: Sources/Sentry/SentryClient.m
Comment (bruno-garcia) [reply Sources/Sentry/SentryClient.m:237]: Since Java can also be used on the server and have hundreds or thousands of transactions per second, that was critical there. In here I suggested we just extend Transaction (like it was done on @HazAT 's SDKs like JS) since it's a client based transaction and also since it's simpler to implement (splitting things have turned out to be a huge head ache on both Java and .NET even). Since no server side SDK here we have less reason to do it so to move faster I'd vote for inhertiting from Event to add the new fields.

Also, captureEvent is what's used on all the test infrastructure. So the captureEnvelope before was hard to test in this repo, but the addded benefit to get all the device context was an added bonus,

**Review by bruno-garcia — COMMENTED** [review]
I see some tests already. Is this still draft?
What's missing to get merged other than address open review comments?

File: Sources/Sentry/Public/SentryHub.h
Comment (bruno-garcia) [reply Sources/Sentry/Public/SentryHub.h:5]: We tested it and it indeed conflicted. So if someone adds an OpenTelemetry library later it would not compile. This is why we have pretty much everything here prefixed with `Sentry`. I'll resolve this with this reason, feel free to reopen if strong disagree.

File: Sources/Sentry/Transactions/SentrySpanContext.h
Comment (bruno-garcia) [reply Sources/Sentry/Transactions/SentrySpanContext.h:15]: I believe, like the previous comment must keep the prefix `Sentry`. Another one I'll close but against feel free to reopen if strong disagree. I pinged @mitsuhiko to confirm on this but afaik these types need to be prefixed. @HazAT maybe can chime in

File: Sources/Sentry/Transactions/SentrySpanId.h
Comment (bruno-garcia) [inline Sources/Sentry/Transactions/SentrySpanId.h:27]: Also seems wrong

**Comment by brustolin** [comment]
Well, now I think I have addressed all comments. Still figuring out this way of using GitHub.
A couple of discussion the be solved yet, but I think is ready to be merged, and let's solve these bigger issues in future PRs.

**Review by bruno-garcia — COMMENTED** [review]
Lets just address the last open issues pending replies from folks in Vienna but other than those points, LGTM

File: Sources/Sentry/Public/SentryClient.h
Comment (marandaneto) [reply Sources/Sentry/Public/SentryClient.h:44]: well, every other method does, `captureEvent` requires a Scope, and right now `captureTransaction` is instantiating a new Scope every time, so either we take a Scope param or pass null, so we save the allocation if not using it at all.

File: Sources/Sentry/Public/SentryHub.h
Comment (marandaneto) [reply Sources/Sentry/Public/SentryHub.h:5]: that makes sense, actually, we had a bug with KSCrash about that, good catch

File: Sources/Sentry/SentryClient.m
Comment (marandaneto) [reply Sources/Sentry/SentryClient.m:237]: > splitting things have turned out to be a huge head ache on both Java and .NET even

I'm not aware of that, why's that?

This will for sure generate a new source of issues here and/or Android, it was decided not to do it on Android to be cheap, either on payload or memory footprint, now we re doing the opposite here for no reason, Mobile has limited resources and will offer if used heavily, think about every transaction doing IO to read available storage, memory, etc... I'm against it till we decide to read only the min. and useful data to be sent over with transactions, your call.
On the Android side, issues will come, where is the device context? why do I have it on iOS and not on Android? and the explanation will be actually the opposite.

Comment (bruno-garcia) [reply Sources/Sentry/SentryClient.m:237]: > > splitting things have turned out to be a huge head ache on both Java and .NET even
> 
> I'm not aware of that, why's that?
> 

Without a clear guidence of what should be set to both transaction and events it happened on both .NET and Java that neither added critical data like environment, release, tags. A lot more code was put into splitting things, to deal with mispatches such as TransactionName vs Transaction (string), and it's a diviation from the current spec that is just unknown territory. This had startTimestamp already on the event since v5 so it was already trying to prepare to have Event match both Transaction and Event.

> This will for sure generate a new source of issues here and/or Android, it was decided not to do it on Android to be cheap, either on payload or memory footprint, now we re doing the opposite here for no reason, Mobile has limited resources and will suffer if used heavily, think about every transaction doing IO to read available storage, memory, etc... I'm against it till we decide to read only the min. and useful data to be sent over with transactions, your call.
> On the Android side, issues will come, where is the device context? why do I have it on iOS and not on Android? and the explanation will be actually the opposite.

>  Android to be cheap

It was decided on Spring to be cheap. Android will have a transaction every few seconds or minutes, compared to hundreds of transactions per second on Spring. So the reason is last important here IMHO

File: Sources/Sentry/Public/SentryClient.h
Comment (bruno-garcia) [reply Sources/Sentry/Public/SentryClient.h:44]: > well, every other method does,

It's not really a reason

> captureEvent requires a Scope, and right now captureTransaction is instantiating a new Scope every time,

This is a good point.

@brustolin Can we pass `nil` instead of creating a scope if we're not using it?

> so either we take a Scope param or pass null, so we save the allocation if not using it at all.

Yes, this ^ please

**Review by bruno-garcia — APPROVED** [review]
@brustolin I believe we agreed on one point, pass `nil` as `scope`. Other than that I think we can merge for now. The Transaction/Event discussion we can address later at least, lets not block this PR on that

File: Sources/Sentry/Public/SentryClient.h
Comment (brustolin) [reply Sources/Sentry/Public/SentryClient.h:44]: For now, it does not accept null in the parameter, and it uses the parameter in a ton of other calls. I guess I could change it all until the sendEvent, but I would have to add some extra flow to verify if scope is nil.

Comment (bruno-garcia) [reply Sources/Sentry/Public/SentryClient.h:44]: OK lets merge this and follow up with that on a next PR.

**Pull request merged**
